net/portmapper/portmapper.go

```diff
@@ -0,0 +1,601 @@
+// their wifi), but we can make sure we invalidate mappings for later when the network
+// comes back.
+func (c *Client) NoteNetworkDown() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.invalidateMappingsLocked(false)
+}
+
+func (c *Client) Close() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.closed {
+		return nil
+	}
+	c.closed = true
+
+	// TODO: close some future ever-listening UDP socket(s),
+	// waiting for multicast announcements from router.
+	return nil
+}
+
+// SetLocalPort updates the local port number to which we want to port
+// map UDP traffic.
+func (c *Client) SetLocalPort(localPort uint16) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.localPort == localPort {
+		return
+	}
+	c.localPort = localPort
+	c.invalidateMappingsLocked(true)
+}
+
+func (c *Client) gatewayAndSelfIP() (gw, myIP netaddr.IP, ok bool) {
+	gw, myIP, ok = interfaces.LikelyHomeRouterIP()
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if !ok {
+		gw = netaddr.IP{}
+		myIP = netaddr.IP{}
+	}
+	if gw != c.lastGW || myIP != c.lastMyIP || !ok {
+		c.lastMyIP = myIP
+		c.lastGW = gw
+		c.invalidateMappingsLocked(true)
+	}
+	return
+}
+
+func (c *Client) invalidateMappingsLocked(releaseOld bool) {
+	if c.pmpMapping != nil {
+		if releaseOld {
+			c.pmpMapping.release()
+		}
+		c.pmpMapping = nil
+	}
+	c.pmpPubIP = netaddr.IP{}
+	c.pmpPubIPTime = time.Time{}
+	c.pcpSawTime = time.Time{}
+	c.uPnPSawTime = time.Time{}
+}
+
+func (c *Client) sawPMPRecently() bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.sawPMPRecentlyLocked()
+}
+
+func (c *Client) sawPMPRecentlyLocked() bool {
+	return !c.pmpPubIP.IsZero() && c.pmpPubIPTime.After(time.Now().Add(-trustServiceStillAvailableDuration))
+}
+
+func (c *Client) sawPCPRecently() bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.pcpSawTime.After(time.Now().Add(-trustServiceStillAvailableDuration))
+}
+
+func (c *Client) sawUPnPRecently() bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.uPnPSawTime.After(time.Now().Add(-trustServiceStillAvailableDuration))
+}
+
+func closeCloserOnContextDone(ctx context.Context, c io.Closer) (stop func()) {
+	// Close uc on ctx being done.
+	ctxDone := ctx.Done()
+	if ctxDone == nil {
+		return func() {}
+	}
+	stopWaitDone := make(chan struct{})
+	go func() {
+		select {
+		case <-stopWaitDone:
+		case <-ctxDone:
+			c.Close()
+		}
+	}()
+	return func() { close(stopWaitDone) }
+}
+
+// NoMappingError is returned by CreateOrGetMapping when no NAT
+// mapping could be returned.
+type NoMappingError struct {
+	Err error
+}
+
+func (nme NoMappingError) Unwrap() error { return nme.Err }
+func (nme NoMappingError) Error() string { return fmt.Sprintf("no NAT mapping available: %v", nme.Err) }
+
+func IsNoMappingError(err error) bool {
+	_, ok := err.(NoMappingError)
+	return ok
+}
+
+var (
+	ErrNoPortMappingServices = errors.New("no port mapping services were found")
+	ErrGatewayDetected       = errors.New("failed to look gateway address")
```

Comment:
failed to look up

The error name doesn't parse for me. ErrGatewayNotFound maybe?